pkg/apitypes/managed_tx.go

```diff
@@ -125,3 +153,37 @@ type TransactionUpdateReply struct {
 	ProtocolID      string       `json:"protocolId"`
 	TransactionHash string       `json:"transactionHash,omitempty"`
 }
+
+// Transaction sub-status entries can be added for a given transaction so a caller
+// can see discrete steps in a transaction moving to confirmation on the blockchain.
+// There only exists a single entry in the list for each unique sub-status type. If
+// a transaction goes through the same sub-status more than once (for example if it
+// has gas recalculated for it more than once) then the "Count" and "LastOccurence" fields
+// should be updated accordingly. This design allows a caller to see when the most recent
+// sub-status changes took place and if necessary, to order them by time, but prevents
+// the potential for the sub-status list to grow indefinitely.
+func (mtx *ManagedTX) AddSubStatus(ctx context.Context, subStatus TxSubStatus) {
+	// See if this status exists in the list already
+	for _, entry := range mtx.SubStatusHistory {
+		if entry.Status == subStatus {
```

Comment:
There seems to be a split approach here that I can't reconcile.

The code below suggests we are aiming to limit the number of unique entries to 50 (I guess the most recent 50?). This fits with an approach where (like k8s `describe`) we de-duplicate when something is in exactly the same state, but try to provide an accurate a history as possible.

However, the comparison on this line only uses the `TxSubStatus` line for comparison bare - without all the interesting things that might make it unique, such as the TX Hash. So it's impossible to get more than 6 entries in the table. In fact the entries don't really contain any detail at all now.

I think we need to choose whether to give detailed information, or not. I believe we should give detailed information - as we do before this PR.

I assert the following are a must, which certainly does introduce the potential for more than 50 records:
- The current Hash of the transaction we are tracking
- The receipt `ProtocolID`

The number of confirmations is slightly interesting, as we could choose to update that on an existing record, rather than produce a new record.

I like the approach you've got of adding the raw information on the JSON, and allowing translation of that into human readable form by the consumer of the API (such as a UX, or application).

So I think I'm proposing:

```go
type TxSubStatusEntry struct {
	Occurrences     int              `json:"occurrences"`        // updates every time we add a duplicate
	FirstOccurrence *fftypes.FFTime  `json:"first"`              // set when first added to array
	LastOccurrence  *fftypes.FFTime  `json:"last"`               // set when first added, and updated on each duplicate
	Status          TxSubStatus      `json:"status"`             // part of uniqueness
	TX              string           `json:"tx"`                 // part of uniqueness
	Receipt         string           `json:"receipt"`            // part of uniqueness
	GasPrice        *fftypes.JSONAny `json:"gasPrice,omitempty"` // part of uniqueness - only present on Status==RetrievedGasPrice
	Confirmations   int              `json:"confirmations"`      // updates as more are added
}

func (s *TxSubStatusEntry) Key() string {
	return fmt.Sprintf("%s:%s:%s:%s", s.Status, s.TX, s.Receipt, s.GasPrice.String())
}
```

Maybe a higher level point as to why I feel quite strongly about ^^^: if the system is looping through fundamentally different sub-step, then I think it's really important that the information shows this.

e.g. I think between (A) and (B) I feel like people need (B):

### Information preservation A

- ✅ validated (1)
- ✅ estimatedgas (2)
- ✅ submitted (3)

### Information preservation B

- ✅ validated (1)
- ✅ estimatedgas (1) GasPrice=12345
- ✅ submitted (1) TXHash=12345
- ✅ estimatedgas (1) GasPrice=12999
- ✅ submitted (2) TXHash=23456

Reply:
... there's a subtle point here too, that I just realized I'm making.

I think the list of sub-status's is a little inconsistent. For some steps we've got before+after states, for others we've just got after states. I think we should be very consistent on that.

The example above is we just "after" states (I didn't include any "before" states)

Reply:
oh, darn - and that suddenly reminded me there might be an error that we're getting preventing us moving to the next state, so I guess there's a decision as to whether we record errors in special "before" states (doubling the length of the array)... or by just updating the previous state with the error preventing us from progressing.

Reply:
... which maybe means we should only have "before" states, rather than only have "after" states:

### Information preservation C

- ✅ estimatingGas (3) LastError="Node unavaialable", LastErrorTime=10:57
- ✅ submitting (1) TXHash=12345,  GasPrice=12345
- ✅ estimatingGas (1)
- ✅ submitted (2) TXHash=23456, GasPrice=12999